### ISSUE TITLE
`terraform_unused_required_providers` errors with implicit module passing

### ISSUE THREAD

**Issue by tmatilai** [terraform-linters/tflint#1328]
#1225 fixed the case where a provider was explicitly passed to a module, but stated:

> It remains a lint error to declare a required provider when it is implicitly inherited by the child module, as that declaration is the child's responsibility.

But this in conflict with the [Terraform best practices](https://www.terraform.io/language/expressions/version-constraints#best-practices):

> **Terraform Core and Provider Versions**
> 
> - Reusable modules should constrain only their minimum allowed versions of Terraform and providers, such as `>= 0.12.0`. This helps avoid known incompatibilities, while allowing the user of the module flexibility to upgrade to newer versions of Terraform without altering the module.
>
> - Root modules should use a `~>` constraint to set both a lower and upper bound on versions for each provider they depend on.

I.e. the root modules should specify a (stricter) version requirement. But `tflint` complains about it.

**Comment by bendrucker** [comment]
Yes, I've also run into this. Your notes/citations are accurate, this should be allowed. Short of disabling the rule entirely, it seems like walking the module tree and accumulating all required providers will be necessary.

This is only an issue when you use environment variables to configure your providers. When a `provider` block is used that will serve as a usage of the provider in the root module.

**Comment by tmatilai** [comment]
> This is only an issue when you use environment variables to configure your providers.

And not all providers even need configuration. For example `null` or `random`.
